File: ops/jobs/src/ugo_jobs/house_docs.py
"""I documenti di casa, indicizzati dal sogno (backlog gruppo 7, ADR-111).

«UGO conosce solo cio che ha sentito»: il libretto della caldaia e il
contratto della luce esistono, e finche non glieli si legge ad alta voce lui
non li sa. La macchina per saperli e gia scritta — bucket privato, estrazione,
finestre, embedding locale, ciphertext — ma vive tutta dentro la reception,
per i clienti (ADR-054). Qui la stessa macchina serve la casa.

Le differenze rispetto a `customer_docs`, e sono due sole:

- lo scope e la **casa** e non il cliente, quindi il muro RLS e un altro e le
  righe stanno in tabelle proprie (`house_documents`, `house_chunks`);
- non c'e nessuna `knowledge_epoch` da far avanzare: la cache delle risposte e
  della reception, e in casa non c'e.
"""

from __future__ import annotations

import json
import os

import psycopg

from .config import JobsConfig
from .crypto import decrypt_text, encrypt_text, parse_data_key
from .customer_chunks import EMBED_BATCH, chunk_text, max_chunks
from .customer_docs import _extract_text, _s3_client
from .embeddings import embed


def house_bucket(cfg: JobsConfig) -> str:
    """Un bucket suo: i documenti di famiglia non stanno con quelli dei clienti."""
    return os.environ.get("S3_BUCKET_HOUSE_DOCS", "ugo-house-docs")


def _replace_chunks(
    conn: psycopg.Connection,
    cfg: JobsConfig,
    *,
    account_id: str,
    document_id: str,
    pieces: list[tuple[str, str]],
) -> int:
    key = parse_data_key(cfg.data_key_b64)
    capped = pieces[: max_chunks()]
    if len(capped) < len(pieces):
        # niente tetti in silenzio (CLAUDE.md): il taglio si stampa
        print(
            json.dumps(
                {"house_chunks_capped": {"document": document_id, "dropped": len(pieces) - len(capped)}}
            ),
            flush=True,
        )
    conn.execute("delete from house_chunks where document_id = %s", (document_id,))
    for start in range(0, len(capped), EMBED_BATCH):
        batch = capped[start : start + EMBED_BATCH]
        # il `ref` viaggia DENTRO l'embedding e resta in chiaro accanto: e un
        # identificatore (pagina, finestra), non contenuto
        vectors = embed(cfg, [f"{ref}\n{piece}" for ref, piece in batch])
        for (ref, piece), vector in zip(batch, vectors, strict=True):
            conn.execute(
                """
                insert into house_chunks (account_id, document_id, ref, text, embedding)
                values (%s, %s, %s, %s, %s)
                """,
                (account_id, document_id, ref, encrypt_text(piece, key), json.dumps(vector)),
            )
    return len(capped)


def sync_document(conn: psycopg.Connection, cfg: JobsConfig, document: dict) -> dict:
    key = parse_data_key(cfg.data_key_b64)
    try:
        payload = _s3_client(cfg).get_object(
            Bucket=house_bucket(cfg), Key=document["s3_key"]
        )["Body"].read()
        text = _extract_text(document["mime"], payload)
        filename = decrypt_text(document["filename"], key)
        written = _replace_chunks(
            conn,
            cfg,
            account_id=str(document["account_id"]),
            document_id=str(document["id"]),
            pieces=chunk_text(text, filename),
        )
        conn.execute(
            "update house_documents set status = 'ok', last_error = null, "
            "indexed_at = now() where id = %s",
            (str(document["id"]),),
        )
        return {"document": str(document["id"]), "chunks": written}
    except Exception as error:  # noqa: BLE001 - un file rotto non ferma il giro
        conn.execute(
            "update house_documents set status = 'error', last_error = %s where id = %s",
            ("extract_failed", str(document["id"])),
        )
        return {"document": str(document["id"]), "error": type(error).__name__}


def run_house_docs(conn: psycopg.Connection, cfg: JobsConfig, account_id: str) -> list[dict]:
    rows = conn.execute(
        """
        select id, account_id, s3_key, filename, mime
        from house_documents
        where account_id = %s and indexed_at is null
        order by uploaded_at
        """,
        (account_id,),
    ).fetchall()
    columns = ["id", "account_id", "s3_key", "filename", "mime"]
    reports = []
    for row in rows:
        reports.append(sync_document(conn, cfg, dict(zip(columns, row, strict=True))))
        conn.commit()
    if reports:
        print(json.dumps({"house_docs": reports}), flush=True)
    return reports

File: packages/db/src/schema/house-docs.ts
import { sql } from "drizzle-orm";
import { index, integer, pgTable, text, timestamp, uuid, vector } from "drizzle-orm/pg-core";
import { EMBEDDING_DIMENSIONS } from "@ugo/shared";
import { customerSourceStatus } from "./enums.js";
import { accountId } from "./accounts.js";

/**
 * I documenti di casa (backlog gruppo 7, ADR-111).
 *
 * «UGO conosce solo ciò che ha sentito»: il libretto della caldaia, il
 * contratto della luce, il referto del veterinario esistono, e finché non
 * glieli si legge ad alta voce lui non li sa. La macchina per saperli è già
 * scritta — bucket privato, estrazione, finestre, embedding locale,
 * ciphertext, ricerca vettoriale — ma vive tutta dentro la reception, per i
 * clienti (ADR-054). Qui la stessa macchina serve la **casa**.
 *
 * Due tabelle e non una: il documento è la cosa che si carica e si butta, il
 * frammento è la cosa che si cerca. Tenerli insieme vorrebbe dire ricalcolare
 * gli embedding per rinominare un file.
 *
 * Perché NON si è riusata `customer_chunks` allargandola: là dentro lo scope è
 * `customer_id` e la reception ci legge sopra. Un documento di famiglia in
 * quella tabella sarebbe a un `where` di distanza dall'occhio di un cliente, e
 * quel `where` sarebbe l'unica cosa fra il referto del veterinario e uno
 * sconosciuto. Due tabelle, due muri.
 */

export const houseDocuments = pgTable(
  "house_documents",
  {
    id: uuid("id")
      .primaryKey()
      .default(sql`gen_random_uuid()`),
    accountId: accountId(),
    /** chiave opaca nel bucket privato: mai il nome del file */
    s3Key: text("s3_key").notNull(),
    /** ciphertext AES-256-GCM: un nome di file è contenuto (regola 6) */
    filename: text("filename").notNull(),
    mime: text("mime").notNull(),
    sizeBytes: integer("size_bytes").notNull().default(0),
    uploadedAt: timestamp("uploaded_at", { withTimezone: true }).notNull().defaultNow(),
    indexedAt: timestamp("indexed_at", { withTimezone: true }),
    status: customerSourceStatus("status").notNull().default("pending"),
    lastError: text("last_error"),
  },
  (table) => [index("house_documents_account_idx").on(table.accountId)],
);

export const houseChunks = pgTable(
  "house_chunks",
  {
    id: uuid("id")
      .primaryKey()
      .default(sql`gen_random_uuid()`),
    accountId: accountId(),
    documentId: uuid("document_id").notNull(),
    /** pagina o finestra: un identificatore, non contenuto */
    ref: text("ref").notNull(),
    /** ciphertext AES-256-GCM sotto la chiave di casa (regola 6) */
    text: text("text").notNull(),
    embedding: vector("embedding", { dimensions: EMBEDDING_DIMENSIONS }).notNull(),
    createdAt: timestamp("created_at", { withTimezone: true }).notNull().defaultNow(),
  },
  (table) => [
    index("house_chunks_document_idx").on(table.documentId),
    index("house_chunks_embedding_hnsw_idx").using(
      "hnsw",
      table.embedding.op("vector_cosine_ops"),
    ),
  ],
);
